code/scripting/api/libs/mission.cpp

```diff
@@ -1031,6 +1032,88 @@ ADE_FUNC(createShip,
 		return ade_set_error(L, "o", l_Ship.Set(object_h()));
 }
 
+ADE_FUNC(createDebris,
+	l_Mission,
+	"[model, submodel, vector position, orientation, enumeration create_flags, number hitpoints, number spark_timeout_seconds, team, shipclass source_class, ship source_ship, vector explosion_center, number explosion_force_multiplier]",
+	"Creates a chunk or shard of debris with the specified parameters.  Vectors are in world coordinates.  Any parameter can be nil or negative to specify defaults.  Create flags can be any combination of DC_IS_HULL, DC_VAPORIZE, DC_IS_PERSISTENT, DC_SET_VELOCITY, or DC_FIRE_HOOK.",
```

Comment:
This API is strange.
What is model, if I want to make a debris piece of a shipclass?
What actually happens if I create Debris of an actual ship instance? Will that die?

I'd expect the API to actually be two functions.
For Hull Debris: (position, orientation, ship_class, submodel / submodel_name, create_flags, hitpoints, spark_timeout_seconds, explosion_center, explosion_force_multiplier)
For non-hull debris: (position, orientation, submodel (since that contains a polymodel pointer), create_flags, hitpoints, spark_timeout_seconds, explosion_center, explosion_force_multiplier)

In addition "to specify defaults" is a problem. What are the defaults? For model? No idea. For position? No idea. For orientation? Probably random or identity (I know from reading the code that its random, but a user will not). Same with most of the other arguments

Reply:
Yeah it's not the best, but I'm not sure how to meaningfully make it better.

Model is there because the code allows you to specify a model but not a submodel if you're creating generic non-hull debris.  Creating debris of a live ship does not cause the live ship to die.

I could explicitly list the defaults of each parameter, but that might get unwieldy.

Reply:
Okay, I suggest the following API for less confusion:
``createDebris(ship | shipclass | model | submodel source, submodel_name | nil submodel_source (must be nil if source is submodel, can be nil to spawn all debris submodel), vector position, orientation, enumeration create_flags, number hitpoints, number spark_timeout_seconds, team, vector explosion_center, number explosion_force_multiplier)``

This has the advantage, that the ship/model/submodel arguments are all mutually exlusive, as makes sense logically here, and the user won't need a bunch of nils in normal operations for the parameters of possible sources the user doesn't need.

To differentiate between the userdata types, you will need a new ade function. I suggest to use these in LuaConvert.h/.cpp:
```cpp
template<typename T>
bool ade_odata_is_userdata_type(lua_State* L, int stackposition, const ::scripting::ade_obj<T>& obj_type) {
	return ade_odata_is_userdata_type(L, stackposition, obj_type.LibIdx);
}

bool ade_odata_is_userdata_type(lua_State* L, int stackposition, size_t typeIdx) {
	// WMC - Get metatable
	lua_getmetatable(L, stackposition);
	int mtb_ldx = lua_gettop(L);
	Assert(!lua_isnil(L, -1));

	// Get ID
	lua_pushstring(L, "__adeid");
	lua_rawget(L, mtb_ldx);

	if (lua_tonumber(L, -1) != typeIdx) {
		lua_pushstring(L, "__adederivid");
		lua_rawget(L, mtb_ldx);
		if ((size_t)lua_tointeger(L, -1) != typeIdx) {
			return false;
		}
		lua_pop(L, 1);
	}
	lua_pop(L, 2);
	return true;
}
```

The logical consequence is also, that the following function is replaced in the same file, with the following code:
```cpp
bool ade_odata_helper(lua_State* L, int stackposition, size_t idx) {
	if (!ade_odata_is_userdata_type(L, stackposition, idx)) {
		// Issue the LuaError here since this is the only place where we have all relevant information
		LuaError(L, "Argument %d is the wrong type of userdata; '%s' given, but '%s' expected", stackposition,
			::scripting::internal::getTableEntry((size_t)lua_tointeger(L, -2)).Name,
			::scripting::internal::getTableEntry(idx).GetName());
		return false;
	}
	else {
		return true;
	}
}
```

Then you can call ``ade_odata_is_userdata_type`` like so, to test for example if the first argument is of type ship:
```cpp
//Stack position is 1-indexed for arguments of called functions
bool is_ship = ade_odata_is_userdata_type(L, 1, l_Ship);
```